core/ArchiveProcessor/RecordBuilder.php

```diff
@@ -0,0 +1,224 @@
+    {
+        $this->maxRowsInTable = $maxRowsInTable;
+        $this->maxRowsInSubtable = $maxRowsInSubtable;
+        $this->columnToSortByBeforeTruncation = $columnToSortByBeforeTruncation;
+        $this->columnAggregationOps = $columnAggregationOps;
+    }
+
+    public function isEnabled()
+    {
+        return true;
+    }
+
+    public function build(ArchiveProcessor $archiveProcessor)
+    {
+        if (!$this->isEnabled()) {
+            return;
+        }
+
+        $numericRecords = [];
+
+        $records = $this->aggregate($archiveProcessor);
+        foreach ($records as $recordName => $recordValue) {
+            if ($recordValue instanceof DataTable) {
+                $this->insertRecord($archiveProcessor, $recordName, $recordValue);
+
+                Common::destroy($recordValue);
+                unset($recordValue);
+            } else {
+                // collect numeric records so we can insert them all at once
+                $numericRecords[$recordName] = $recordValue;
+            }
+        }
+        unset($records);
+
+        if (!empty($numericRecords)) {
+            $archiveProcessor->insertNumericRecords($numericRecords);
+        }
+    }
+
+    public function buildMultiplePeriod(ArchiveProcessor $archiveProcessor)
+    {
+        if (!$this->isEnabled()) {
+            return;
+        }
+
+        $requestedReports = $archiveProcessor->getParams()->getArchiveOnlyReportAsArray();
+        $foundRequestedReports = $archiveProcessor->getParams()->getFoundRequestedReports();
+
+        $recordsBuilt = $this->getRecordMetadata($archiveProcessor);
+
+        $numericRecords = array_filter($recordsBuilt, function (Record $r) { return $r->getType() == Record::TYPE_NUMERIC; });
+        $blobRecords = array_filter($recordsBuilt, function (Record $r) { return $r->getType() == Record::TYPE_BLOB; });
+
+        // TODO: dependsOn in record metadata
+        foreach ($blobRecords as $record) {
+            if (!empty($requestedReports)
+                && !in_array($record->getName(), $requestedReports)
+                && !in_array($record->getName(), $foundRequestedReports)
+            ) {
+                continue;
+            }
+
+            $maxRowsInTable = $record->getMaxRowsInTable() ?? $this->maxRowsInTable;
+            $maxRowsInSubtable = $record->getMaxRowsInSubtable() ?? $this->maxRowsInSubtable;
+            $columnToSortByBeforeTruncation = $record->getColumnToSortByBeforeTruncation() ?? $this->columnToSortByBeforeTruncation;
+
+            $archiveProcessor->aggregateDataTableRecords(
+                $record->getName(),
+                $maxRowsInTable,
+                $maxRowsInSubtable,
+                $columnToSortByBeforeTruncation,
+                $this->columnAggregationOps
+            );
+        }
+
+        if (!empty($numericRecords)) {
+            $numericMetrics = array_map(function (Record $r) { return $r->getName(); }, $numericRecords);
```

Comment:
fyi here we would be still aggregating all the records at once, not just the ones that were requested. 

Eg  I requested this url: `index.php?date=2022-08-02,2022-12-30&filter_limit=300&flat=1&format=JSON&format_metrics=1&idSite=12&language=en&method=Goals.getMetrics&module=API&period=range&segment=&token_auth=ANONYMOUS&force_api_session=1&translateColumnNames=1&convertToUnicode=0&1&idGoal=27`
It used these records
<img width="996" alt="image" src="https://user-images.githubusercontent.com/273120/225189139-6d68ac4e-4fb2-49d2-878e-9f959ef88f81.png">

<img width="839" alt="image" src="https://user-images.githubusercontent.com/273120/225189183-4af7b3fc-a809-49d1-8563-f998ffea417d.png">

but then here it still aggregated all of them. In my case over 80 of them when only 4 or so would be needed 
<img width="1321" alt="image" src="https://user-images.githubusercontent.com/273120/225190363-b45aca15-076e-4773-9e4b-5760ff4e9704.png">

Not sure if we can maybe also skip records here when they aren't requested?

Reply:
:+1: will look into it

Reply:
looks like this was caused by a typo